pkg/kv/kvserver/store_raft.go

```diff
@@ -34,32 +34,108 @@ type raftRequestInfo struct {
 }
 
 type raftReceiveQueue struct {
-	syncutil.Mutex
-	infos []raftRequestInfo
+	mu struct { // not to be locked directly
+		syncutil.Mutex
+		infos []raftRequestInfo
+	}
+}
+
+// Len returns the number of requests in the queue.
+func (q *raftReceiveQueue) Len() int {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return len(q.mu.infos)
 }
 
-func (q *raftReceiveQueue) drain() ([]raftRequestInfo, bool) {
-	q.Lock()
-	defer q.Unlock()
-	if len(q.infos) == 0 {
+// Drain moves the stored requests out of the queue, returning them to
+// the caller. Returns true if the returned slice was not empty.
+func (q *raftReceiveQueue) Drain() ([]raftRequestInfo, bool) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.drainLocked()
+}
+
+func (q *raftReceiveQueue) drainLocked() ([]raftRequestInfo, bool) {
+	if len(q.mu.infos) == 0 {
 		return nil, false
 	}
-	infos := q.infos
-	q.infos = nil
+	infos := q.mu.infos
+	q.mu.infos = nil
 	return infos, true
 }
 
-func (q *raftReceiveQueue) recycle(processed []raftRequestInfo) {
+// Recycle makes a slice that the caller knows will no longer be accessed
+// available for reuse.
+func (q *raftReceiveQueue) Recycle(processed []raftRequestInfo) {
 	if cap(processed) > 4 {
 		return // cap recycled slice lengths
 	}
-	q.Lock()
-	defer q.Unlock()
-	if q.infos == nil {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	if q.mu.infos == nil {
 		for i := range processed {
 			processed[i] = raftRequestInfo{}
 		}
-		q.infos = processed[:0]
+		q.mu.infos = processed[:0]
+	}
+}
+
+func (q *raftReceiveQueue) Append(
+	req *kvserverpb.RaftMessageRequest, s RaftMessageResponseStream,
+) (shouldQueue, appended bool) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	if len(q.mu.infos) >= replicaRequestQueueSize {
+		return false, false
+	}
+	q.mu.infos = append(q.mu.infos, raftRequestInfo{
+		req:        req,
+		respStream: s,
+	})
+	// The operation that enqueues the first message will
+	// be put in charge of triggering a drain of the queue.
+	return len(q.mu.infos) == 1, true
+}
+
+type raftReceiveQueues struct {
+	m syncutil.IntMap // RangeID -> *raftReceiveQueue
+	// delMu is held while deleting an entry from `m`. This
+	// is necessary since `IntMap` does not return the element
+	// being removed (if any) and we want stronger synchronization
+	// to prevent memory accounting leaks.
+	delMu syncutil.Mutex
```

Comment:
Can we not make `IntMap.Delete()` return the removed entry instead?

Reply:
`IntMap.Delete` is a line-by-line port of `sync.Map` and I don't want to mess with it.

Reply:
I think I can remove `delMu` again though, in the next PR. Then we will be able to mark individual map entries as deleted, so we can do that first, then delete the entry.

Reply:
What about updating it with `sync.Map.LoadAndDelete` then?

https://pkg.go.dev/sync#Map.LoadAndDelete

nbd either way.